### PR TITLE
Add Recipe doc for implementing AMP on a site

### PR DESCRIPTION
**Pull request by kaedroho**
This document describes a method for hosting an AMP version of a Wagtail site on a separate prefix to the rest of the site and how to change the image tags on that site to use ``<amp-img>`` instead of ````.

This was sponsored by The Motley Fool

### CODE REVIEW

File: docs/advanced_topics/images/amp.rst [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+            attrs.update(extra_attributes)
+
+            if amp_mode_active():
+                return mark_safe('<amp-img{}>'.format(flatatt(attrs)))
+            else:
+                return mark_safe('<img{}>'.format(flatatt(attrs)))
+
+        ...
+
+Without a custom image model, you will have to monkey-patchi the builtin
```
Comment (zerolab) [inline docs/advanced_topics/images/amp.rst]: `monkey-patchi` → `monkey patch` ?

File: docs/advanced_topics/images/amp.rst [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                'context_processors': [
+                    ...
+                    # Add this after other context processors
+                'myapp.amp_context_processors.amp',
+                ],
+            },
+        },
+    ]
+
+You should now be able to use the ``amp_mode_active`` variable in templates.
```
Comment (zerolab) [inline docs/advanced_topics/images/amp.rst]: Does this need an example on how to vary your template based on the amp mode?

**Comment by tomdyson** [comment]
Should we link to Parbhat's articles too? 

https://parbhatpuri.com/amplify-wagtail-django-site-urls-part-1.html
https://parbhatpuri.com/amplify-wagtail-django-site-validation-part-2

File: docs/advanced_topics/images/amp.rst [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+To make changes, we need to add a way to detect which URL was used to render
+the page. To do this, we will have to wrap Wagtail's ``serve()`` view and
+set a thread local to indicate to all downstream code that AMP mode is active.
+
+.. note:: Why a thread local?
+
+    (feel free to skip this part if you're not interested)
+
+    Modifying the ``request`` would obviously be the easiest way to do this.
+    However, the image tag rendering is performed in a part of Wagtail that
```
Comment (cyface) [inline docs/advanced_topics/images/amp.rst]: Ah ha!

File: docs/advanced_topics/images/amp.rst [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        # Change this line to point at your amp_urls instead of Wagtail's urls
+        url(r'amp/', include(wagtail_amp_urls)),
+
+        url(r'', include(wagtail_urls)),
+    ]
+
+After this, there shouldn't be any noticable difference to the AMP version of
+the site.
+
+Write a template context process so that AMP state can be checked in templates
```
Comment (cyface) [inline docs/advanced_topics/images/amp.rst]: processor

File: docs/advanced_topics/images/amp.rst [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+---------------------------------------------------------------
+
+Finally, let's change Wagtail's ``{% image %}`` tag so it renders an ``<amp-img>``
+tags when rendering pages with AMP enabled. We'll make the change on the
+`Rendition` model itself so it applies to both images rendered with the
+``{% image %}`` tag and images rendered in rich text fields as well.
+
+Doing this with a :ref:`Custom image model <custom_image_model>` is easier, as
+you can override the ``img_tag``  method on your custom ``Rendition`` model to
+return a different tag.
```
Comment (cyface) [inline docs/advanced_topics/images/amp.rst]: ah ha!

File: docs/advanced_topics/images/amp.rst (hunk 2 above)
Comment (kaedroho) [reply docs/advanced_topics/images/amp.rst]: This is covered below, did you mean add a more realistic example?

Comment (zerolab) [reply docs/advanced_topics/images/amp.rst]: Yes.
My other thought was that lots of people would use separate templates all together. And that could be either includes in the template, or vary the template in the view/ serve method. But as I write it I realize that is one of those things that we should not be prescriptive about and the ` {% if amp_mode_active %}{% endif %}` example below is good enough

Comment (kaedroho) [reply docs/advanced_topics/images/amp.rst]: Added a section about this. I've also moved it to the "Advanced topics" root as it's no longer specific to images.

File: docs/advanced_topics/images/amp.rst [hunk 6]
```diff
@@ -219,6 +219,65 @@ For example:
         AMP MODE IS ACTIVE!
     {% endif %}
 
+Using a different page template when AMP mode is active
+-------------------------------------------------------
+
+You're probably not going to want to use the same templates on the AMP site as
+you do on the regular web site. So let's add some logic in so Wagtail will
+automatically use separate whenever a page is served with AMP enabled.
```
Comment (zerolab) [inline docs/advanced_topics/images/amp.rst]: Missing "template" in "user separate whenever"

File: docs/advanced_topics/images/amp.rst [hunk 7]
```diff
@@ -219,6 +219,65 @@ For example:
...
+
+We can implement this logic as a mixin which allows us to implement it on many
```
Comment (zerolab) [inline docs/advanced_topics/images/amp.rst]: "We can use a mixin, which allows us to re-use the logic on different page types" ?
At the very least drop the "implement .. implement" construct :)

Comment (kaedroho) [reply docs/advanced_topics/images/amp.rst]: Thanks!

File: docs/advanced_topics/images/amp.rst (hunk 6 above)
Comment (kaedroho) [reply docs/advanced_topics/images/amp.rst]: Fixed, thanks!

**Pull request merged**